middlewares/checkIfUserAllowed.js

```diff
@@ -0,0 +1,73 @@
+const { ROLE_NAME, RESOURCE } = require("../constants/index.js");
+const Comment = require("../models/commentModel");
+const Post = require("../models/postModel.js");
+const User = require("../models/userModel.js");
+
+const checkIfUserAllowed = (resource) => async (req, res, next) => {
+  try {
+    const id = req.params.id;
+    if (resource === RESOURCE.User) {
+      if (
+        req.method === "DELETE" ||
+        ("PUT" &&
+          (req.userRole === ROLE_NAME.ADMIN ||
+            req.userRole === ROLE_NAME.SUPERADMIN))
+      ) {
+        console.log(req.method);
+        console.log(req.userRole);
+        next();
+        return;
+      }
+    }
+    if (
+      req.method === "DELETE" &&
+      (req.userRole === ROLE_NAME.ADMIN ||
+        req.userRole === ROLE_NAME.SUPERADMIN)
+    ) {
+      console.log(req.method);
+      console.log(req.userRole);
+      next();
+      return;
+    }
+
+    let currentResource;
+    switch (resource) {
+      case RESOURCE.User:
+        currentResource = await User.query().findById(id);
+        break;
+      case RESOURCE.UserProfile:
+        currentResource = await UserProfile.query().findById(id);
+
+        break;
+
+      case RESOURCE.Post:
+        currentResource = await Post.query().findById(id);
+
+        break;
+      case RESOURCE.Comment:
+        currentResource = await Comment.query().findById(id);
+        break;
+      default:
```

Comment:
add this kind of message to error

default:
    throw new Error("Resource type is not supported.");